### PR TITLE
Feat/vaults

### PR DESCRIPTION
**Pull request by GGabriele**
Requires https://github.com/Kong/go-kong/pull/224

### CODE REVIEW

**Comment by GGabriele** [comment]
Tested and confirmed this works well with DB-less declarative config too.

File: tests/integration/testdata/sync/012-vaults/kong3x.yaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  routes:
+  - name: r1
+    id: 87b6a97e-f3f7-4c47-857a-7464cb9e202b
+    https_redirect_status_code: 301
+    paths:
+    - /r1
+certificates:
+- id: 13c562a1-191c-4464-9b18-e5222b46035b
+  cert: "{vault://env/my-secret-cert}"
+  key: "{vault://env/my-secret-key}"
```
Comment (aboudreault) [inline tests/integration/testdata/sync/012-vaults/kong3x.yaml:22]: if we want to test the vault **entity**, we should set these vault references to `key: "{vault://my-env-vault/key}`

**Review by aboudreault — APPROVED** [review]
LGTM @GGabriele, only a minor comment in tests. I tested this deck feature with kong gateway 3.0 and it works well (sync, dump, resync diff, etc.). Also tested a vault reference in a plugin (rate-limiting) and things look good there too. :+1:

File: types/postProcess.go [hunk 2]
```diff
@@ -346,3 +346,19 @@ func (crud documentPostAction) Delete(ctx context.Context, args ...crud.Arg) (cr
...
+	return nil, crud.currentState.Vaults.Add(*args[0].(*state.Vault))
+}
+
+func (crud vaultPostAction) Delete(ctx context.Context, args ...crud.Arg) (crud.Arg, error) {
+	return nil, crud.currentState.Vaults.Delete(*((args[0].(*state.Vault)).ID))
+}
+
+func (crud vaultPostAction) Update(ctx context.Context, args ...crud.Arg) (crud.Arg, error) {
+	return nil, crud.currentState.Vaults.Update(*args[0].(*state.Vault))
+}
```
Comment (mikefero) [inline types/postProcess.go:364]: No action needed.

Doesn't look like there are doc strings for these exported functions; however that is the case for the others.

File: types/vault.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+	return &state.Vault{Vault: *updatedVault}, nil
+}
+
+type vaultDiffer struct {
+	kind crud.Kind
+
+	currentState, targetState *state.KongState
+}
+
+func (d *vaultDiffer) Deletes(handler func(crud.Event) error) error {
```
Comment (mikefero) [inline types/vault.go:74]: Should we add a doc string for this exported function?

File: types/vault.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		}, nil
+	}
+	if err != nil {
+		return nil, fmt.Errorf("looking up vault %q: %w",
+			*vault.Prefix, err)
+	}
+	return nil, nil
+}
+
+func (d *vaultDiffer) CreateAndUpdates(handler func(crud.Event) error) error {
```
Comment (mikefero) [inline types/vault.go:112]: Same here for the doc strings.

File: tests/integration/testdata/sync/012-vaults/kong3x.yaml (hunk 1 above)
Comment (GGabriele) [reply tests/integration/testdata/sync/012-vaults/kong3x.yaml:22]: Good catch, thanks!

File: .ci/setup_kong_ee.sh [hunk 5]
```diff
@@ -2,6 +2,61 @@
...
+VZj/GBc9uVunOV5aMcMIs2dFU2fwH4ZjOwnv7rJRldoW1Qlk3uWeIktbmv4yZKvu
+FWPuo3ks+7B+BniqKXuYkNcuhlE+iVr3kJ55qRgX1RxKo4CB3Tynkp7sikp4al7x
+jlHSM9YAqvPFFMOhlU2U3SxE4CLasL99zP0ChINKp9XqzW/qo+F0/Jd4rZmddU2f
+M9Cx62cc0L5IlsHLVJj3zwHZzc/ifpBUeebB98IjoQAfiRkbX0Oe/c2TxtR4o/RH
+GWNeKCThdliZkXaLiOPswOV1BYfA00etorcY0CIy0aTaZgfvrYsJe4aT/hkF/JvF
+tHJ/iD67m8RhLysRL/w50+quVMluUDqJps0HhKrB9wzNJWrddWhvplVeuOXEJfTM
+i80W1JE4OApdISMEn56vRi+BMQMgIeYWznfyQnI4G3rUJQFMI5KzLxkvfYNTF3Ci
+3Am0KaJ7X2oLOq4Qc6CM3qDkvvId61COlfJb2Fo3ETnoT66mxcb6bjtz1WTWOopm
+UcmBKErRUKksINUxwuvP/VW007tXOjZH7wmiM2IW5LUZVkbhB1iE
+-----END CERTIFICATE-----'''
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply .ci/setup_kong_ee.sh:30]: Yeah it doesn't matter much here, the certificates entity doesn't perform any validation

File: file/builder.go [hunk 6]
```diff
@@ -631,6 +631,32 @@ func (b *stateBuilder) routes() {
...
+				b.err = err
+				return
+			} else {
+				v.ID = kong.String(*vault.ID)
+			}
+		}
+		utils.MustMergeTags(&v.Vault, b.selectTags)
+
+		b.rawState.Vaults = append(b.rawState.Vaults,
+			&v.Vault)
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply file/builder.go:658]: Nope!

File: state/vault.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+			panic(unexpectedType)
+		}
+		return &Vault{Vault: *vault.DeepCopy()}, nil
+	}
+	return nil, ErrNotFound
+}
+
+// Get gets a vault by prefix or ID.
+func (k *VaultsCollection) Get(prefixOrID string) (*Vault, error) {
+	if prefixOrID == "" {
```
[the remark this answers is not in the document]
Comment (GGabriele) [reply state/vault.go:85]: Nope, it would result in a schema error.

File: types/vault.go (hunk 4 above)
Comment (GGabriele) [reply types/vault.go:116]: added

**Review by mikefero — DISMISSED** [review]

**Review by aboudreault — APPROVED** [review]

**Pull request merged**